File: 90-subsets-ii/90-subsets-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsetsWithDup = function(nums) {
    
    function subsets(nums, l, r, subset=[]) {
        
        if (l > r) {
            let copy = [...subset];
            let copyStr = copy.join();
            
            if (dp[copyStr]) return;
            else {
                result.push(copy);
                dp[copyStr] = true;
            }
            
            return;
        }
        
        subset.push(nums[l]);
        
        subsets(nums, l+1, r, subset);
        
        subset.pop();
        
        subsets(nums, l+1, r, subset);
    }
    
    var result = [ ]
    var dp = {}
    nums.sort();
    subsets(nums, 0, nums.length - 1);
    
    return result;
};

File: 90-subsets-ii/90-subsets-ii.java
class Solution {
    
    List<List<Integer>> result = new ArrayList<List<Integer>>();
    HashMap<String, Boolean> map = new HashMap<String, Boolean>();
    
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        
        List<Integer> subset = new ArrayList<Integer>();
        
        Arrays.sort(nums);
        
        subsets(nums, 0, nums.length - 1, subset);
        
        return result;
    }
    
    public void subsets(int[] nums, int l, int r, List<Integer> subset) {
        
        if (l > r) {
            List<Integer> copy = new ArrayList<Integer>(subset);
            String copyStr = copy.toString();
            
            if (map.get(copyStr) != null) return;
            
            map.put(copyStr, true);
            
            result.add(copy);
            
            return;
        }
        
        subset.add(nums[l]);
        
        subsets(nums, l+1, r, subset);
        
        subset.remove(subset.size() - 1);
        
        subsets(nums, l+1, r, subset);
    }
}